public/app/features/datasources/state/reducers.ts

```diff
@@ -29,61 +19,75 @@ export const initialState: DataSourcesState = {
   dataSourceMeta: {} as DataSourcePluginMeta,
 };
 
-export const dataSourcesReducer = reducerFactory(initialState)
-  .addMapper({
-    filter: dataSourcesLoaded,
-    mapper: (state, action) => ({
+export const dataSourceLoaded = createAction<DataSourceSettings>('dataSources/dataSourceLoaded');
+export const dataSourcesLoaded = createAction<DataSourceSettings[]>('dataSources/dataSourcesLoaded');
+export const dataSourceMetaLoaded = createAction<DataSourcePluginMeta>('dataSources/dataSourceMetaLoaded');
+export const dataSourcePluginsLoad = createAction('dataSources/dataSourcePluginsLoad');
+export const dataSourcePluginsLoaded = createAction<DataSourceTypesLoadedPayload>(
+  'dataSources/dataSourcePluginsLoaded'
+);
+export const setDataSourcesSearchQuery = createAction<string>('dataSources/setDataSourcesSearchQuery');
+export const setDataSourcesLayoutMode = createAction<LayoutMode>('dataSources/setDataSourcesLayoutMode');
+export const setDataSourceTypeSearchQuery = createAction<string>('dataSources/setDataSourceTypeSearchQuery');
+export const setDataSourceName = createAction<string>('dataSources/setDataSourceName');
+export const setIsDefault = createAction<boolean>('dataSources/setIsDefault');
+
+export const dataSourcesReducer = (state: DataSourcesState = initialState, action: AnyAction): DataSourcesState => {
+  if (dataSourcesLoaded.match(action)) {
+    return {
       ...state,
       hasFetched: true,
       dataSources: action.payload,
       dataSourcesCount: action.payload.length,
-    }),
-  })
-  .addMapper({
-    filter: dataSourceLoaded,
-    mapper: (state, action) => ({ ...state, dataSource: action.payload }),
-  })
-  .addMapper({
-    filter: setDataSourcesSearchQuery,
-    mapper: (state, action) => ({ ...state, searchQuery: action.payload }),
-  })
-  .addMapper({
-    filter: setDataSourcesLayoutMode,
-    mapper: (state, action) => ({ ...state, layoutMode: action.payload }),
-  })
-  .addMapper({
-    filter: dataSourcePluginsLoad,
-    mapper: state => ({ ...state, plugins: [], isLoadingDataSources: true }),
-  })
-  .addMapper({
-    filter: dataSourcePluginsLoaded,
-    mapper: (state, action) => ({
+    };
+  }
+
+  if (dataSourceLoaded.match(action)) {
+    return { ...state, dataSource: action.payload };
+  }
+
+  if (setDataSourcesSearchQuery.match(action)) {
+    return { ...state, searchQuery: action.payload };
+  }
+
+  if (setDataSourcesLayoutMode.match(action)) {
+    return { ...state, layoutMode: action.payload };
+  }
+
+  if (dataSourcePluginsLoad.match(action)) {
+    return { ...state, plugins: [], isLoadingDataSources: true };
+  }
```

Comment:
this reducer looks different from the rest, any reason?

Reply:
I hope this looks the same as the reducers for location, dashboard and explore? This reducer can't be immutable so we need to create the reducer ourselves.

Reply:
Or were you thinking about something else that I've missed?

Reply:
No, just wondering why it did did not look like the others :) but see now that this is another case of a reducer having mutable state